quic/s2n-quic-core/src/recovery/bbr/probe_bw.rs

```diff
@@ -0,0 +1,759 @@
+        let rounds = reno_rounds
+            .try_into()
+            .unwrap_or(u8::MAX)
+            .min(MAX_BW_PROBE_ROUNDS);
+        self.rounds_since_bw_probe >= rounds
+    }
+
+    /// Start the `Cruise` cycle phase
+    fn start_cruise(&mut self) {
+        debug_assert_eq!(self.cycle_phase, CyclePhase::Down);
+
+        self.cycle_phase = CyclePhase::Cruise
+    }
+
+    /// Start the `Up` cycle phase
+    fn start_up(
+        &mut self,
+        round_counter: &mut round::Counter,
+        delivered_bytes: u64,
+        cwnd: u32,
+        max_data_size: u16,
+        now: Timestamp,
+    ) {
+        debug_assert_eq!(self.cycle_phase, CyclePhase::Refill);
+
+        self.bw_probe_samples = true;
+        self.ack_phase = AckPhase::ProbeStarting;
+        round_counter.set_round_end(delivered_bytes);
+        self.cycle_stamp = Some(now);
+        self.cycle_phase = CyclePhase::Up;
+        self.raise_inflight_hi_slope(cwnd, max_data_size);
+    }
+
+    /// Start the `Refill` cycle phase
+    fn start_refill(
+        &mut self,
+        data_volume_model: &mut data_volume::Model,
+        data_rate_model: &mut data_rate::Model,
+        round_counter: &mut round::Counter,
+        delivered_bytes: u64,
+    ) {
+        debug_assert!(
+            self.cycle_phase == CyclePhase::Down || self.cycle_phase == CyclePhase::Cruise
+        );
+
+        data_volume_model.reset_lower_bound();
+        data_rate_model.reset_lower_bound();
+        self.bw_probe_up_rounds = 0;
+        self.bw_probe_up_acks = 0;
+        self.ack_phase = AckPhase::Refilling;
+        round_counter.set_round_end(delivered_bytes);
+        self.cycle_phase = CyclePhase::Refill;
+    }
+
+    /// Start the `Down` cycle phase
+    fn start_down<Rnd: random::Generator>(
+        &mut self,
+        congestion_state: &mut congestion::State,
+        round_counter: &mut round::Counter,
+        delivered_bytes: u64,
+        random_generator: &mut Rnd,
+        now: Timestamp,
+    ) {
+        congestion_state.reset();
+        self.bw_probe_up_cnt = u32::MAX;
+        self.pick_probe_wait(random_generator);
+        self.cycle_stamp = Some(now);
+        self.ack_phase = AckPhase::ProbeStopping;
+        round_counter.set_round_end(delivered_bytes);
+        self.cycle_phase = CyclePhase::Down;
+    }
+
+    fn pick_probe_wait<Rnd: random::Generator>(&mut self, _random_generator: &mut Rnd) {
+        // TODO:
+        //     /* Decide random round-trip bound for wait: */
+        //     BBR.rounds_since_bw_probe = random_int_between(0, 1); /* 0 or 1 */
+        //     /* Decide the random wall clock bound for wait: */
+        //     BBR.bw_probe_wait = 2sec + random_float_between(0.0, 1.0) /* 0..1 sec */
+        self.rounds_since_bw_probe = Counter::default();
+        self.bw_probe_wait = Duration::from_secs(2);
+    }
+}
+
+/// Methods related to the ProbeBW state
+impl BbrCongestionController {
```

Comment:
placing the impl here might make it harder to find and navigate. ill leave it up to you though

Reply:
yeah, my thinking was that there are going to end up being so many methods in BbrCongestionController that I could at least organize them into the separate modules instead of having them all in one long file